wallixbastion 3.0.0: add new commands, return id from created objects…

yoennb opened the pull request:
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
This major release removes deprecated commands/arguments, adds a bunch of new command with support for null arguments and improves workflow by returning the ID of created objects in the command output. Details in the release notes.

## Must have
- [x] Tests
- [x] Documentation

tcarmeli1 commented:
> Thank you for your contribution. Your generosity and caring are unrivaled! Make sure to register your contribution by filling the [Contribution Registration](https://forms.gle/XDfxU4E61ZwEESSMA) form, so our content wizard @tcarmeli1 will know the proposed changes are ready to be reviewed. For your convenience, here is a [link](https://xsoar.pan.dev/docs/contributing/sla) to the contributions SLAs document.

Hi @yoennb, Please fill in the form so I can start reviewing, thanks.

tcarmeli1 commented on Packs/WALLIXBastion/ReleaseNotes/3_0_0.md:
```diff
@@ -0,0 +1,70 @@
+    - wab-get-target-group-restrictions
+    - wab-get-timeframe
+    - wab-get-timeframes
+    - wab-get-user-group-restrictions
```
Please add:
wab-get-version

tcarmeli1 commented on Packs/WALLIXBastion/ReleaseNotes/3_0_0.json:
```diff
@@ -0,0 +1,4 @@
+{
+    "breakingChanges": true,
+    "breakingChangesNotes": null
```
please add the breaking changes explanation inside the breakingChangesNotes field (so it will pop up for the customers as they update the version)

yoennb replied:
Is there any tool to generate it? I saw in the docs that it expects an html string but writing html in json is not the most convenient thing to do

tcarmeli1 commented on Packs/WALLIXBastion/ReleaseNotes/3_0_0.md:
```diff
@@ -0,0 +1,70 @@
+    - wab-get-profile(s): WAB.profile_get.gui_features.dashboards
+    - wab-get-application(s)/wab-get-device(s): and wab-get-profiles: WAB.(...).tags.id
+
+- added corresponding API endpoint URL to the commands' description
```
is it really necessary?

yoennb replied:
The name of the commands is generated from the endpoints description, but this can be slightly confusing sometimes when two resources are closely related. Seeing the actual url which will be called helps ensure we are interacting with the expected resource and helps linking the integration commands to the API documentation available on the Bastion. We think it might be handy in some cases.

tcarmeli1 replied:
Our code standard is to remove this url from the description.
Would appreciate if you rethink about it, in most cases the user is not aware of the endpoint which is being used so it is not relevant for them in our opinion.
Thank you!

tcarmeli1 commented on Packs/WALLIXBastion/Integrations/WAB/WAB.py:
```diff
@@ -11,34 +11,61 @@ class AuthError(Exception):
         return "Success (failed to format output: %s)" % str(e)
 
 
+def build_params(**kwargs):
```
why do we need this wrapper function? it only calls to assign_params?

yoennb replied:
Its purpose is just to automate calls to assign_params with the proper arguments, especially values_to_ignore.

tcarmeli1 replied:
The best practice is to call the assign_params directly.
Would appreciate if yu rethink about it.
Thank you!

tcarmeli1 reviewed (changes requested):
Hi @yoennb ,
Thanks a lot for your contribution!
Amazing work :)
Please see my review below.
Some general comments:
in the WAB.yml:
1. In most commands' description a url is added - is it really needed? I would suggest deleting this one since it is not related to the user.
2. In all arguments that has accepted values/specific options for answers our standard is to use the predefined argument setting.
You should add the following fields (attached an example):
![image](https://github.com/user-attachments/assets/368422a9-c8f7-462f-9c2f-ae01600f559a)
auto: PREDEFINED
predefined:
/- option1
/- option 2
This way, the user will have all options and will be able to choose one easily.
3. I have noticed in some commands the 'q' argument name is used - i would suggest changing the name to a more informative one, but it is only a suggestion and it is for your decision.
Feel free to contact me,
Tal Carmeli

yoennb commented on Packs/WALLIXBastion/Integrations/WAB/WAB.yml:
```diff
@@ -4123,8 +5119,39 @@ script:
+      description: 'The scan id or name to edit.'
+      required: true
+    - name: force
+      description: The default value is false. When it is set to true the values of the subnets, banner_regex or protocols replaced, otherwise the values are added to the existing ones.
```
"are replaced", the verb is indeed missing. will fix

yoennb commented on Packs/WALLIXBastion/Integrations/WAB/WAB.py:
```diff
@@ -11,34 +11,61 @@ class AuthError(Exception):
-    return str(args.get(name, ""))
+Unset = object()
 
 
```
The functions were already here but they have been modified to handle cases were null values are expected and differ from absent values. Since this only apply to specific arguments of specific endpoints, I cannot easily use a general post-processing function, it was simpler to handle it in the argument parsers, with a distinctive value for null vs absent properties

tcarmeli1 replied:
We think it is more elegant to use the original assign_params then convert the null value to None.
Would appreciate if you rethink about it.
Thank you!

tcarmeli1 commented on Packs/WALLIXBastion/Integrations/WAB/WAB.yml:
```diff
@@ -241,8 +229,27 @@ script:
+      description: The account id.
+      required: true
+    - name: credential_type
+      description: '''password'' to change the password or ''ssh_key'' to change the SSH key.'
```
Please use here predefined

tcarmeli1 commented on Packs/WALLIXBastion/Integrations/WAB/WAB.yml:
```diff
@@ -1647,11 +1938,25 @@ script:
+    - name: from_date
+      description: 'Return authentications from this date/time (format: "yyyy-mm-dd" or "yyyy-mm-dd hh:mm:ss").'
+    - name: to_date
+      description: 'Return authentications until this date/time (format: "yyyy-mm-dd" or "yyyy-mm-dd hh:mm:ss").'
```
maybe add a validation for the format inside the code? and return an error if not

yoennb replied:
Unfortunately, this part of the integration is generated. It is easy to patch the yaml afterwards, but it is a bit more complicated for the python code. The Bastion will return an error if the format is not valid. Hope that's acceptable.

tcarmeli1 replied:
If the error message is inappropriate, that's fine with me.

tcarmeli1 commented on Packs/WALLIXBastion/Integrations/WAB/WAB.yml:
```diff
@@ -1647,11 +1938,25 @@ script:
+      description: An authentication id. If specified, only this authentication is returned.
+      required: true
+    - name: from_date
+      description: 'Return authentications from this date/time (format: "yyyy-mm-dd" or "yyyy-mm-dd hh:mm:ss").'
```
maybe add a validation for the format inside the code? and return an error if not

yoennb replied:
same

tcarmeli1 replied:
If the error message is inappropriate, that's fine with me.

tcarmeli1 reviewed (changes requested):
Hi @yoennb ,
**Amazing** work! :)
Thank you for your response.
Please find my comments below, and feel free to reach out if you need any assistance, you can also find me on slack- Tal Carmeli.
After fixing those comments, I’d be glad to schedule a demo.
Since we'll be on holiday until October 4th, please let me know your availability after that date.
Once again, thank you so much!

yoennb commented:
Hello,

I guess you meant October 10th, otherwise I'll have to be very patient :smile: 

I understand your point regarding the url given in the descriptions. If that's fine for you, I replaced it with the API category involved by the command, which sounds more "high-level".

I also reworked how null values are managed as you suggested, directly calling assign_params then fixing nulls in _http_request. Hope you find it better.

Regarding the "q" parameter, I prefer not to change it to avoid possible future collisions, as well as to avoid changing it as well in all pre-existing commands to prevent causing too many breaking changes.

I can also send you the demo in a video if you wish. Otherwise, let's schedule it on slack.

tcarmeli1 commented:
Hi @yoennb ,
Thank you for your response and updates.
Let's schedule a meeting at your earliest convenience. 
I saw the changes you made with the URL in the description—that looks much better!
Regarding the "q" parameter, let's leave it as is.
As for the null handling, thank you for the change. I believe we can also remove the new functionality wrapper for *_arg, can't we?
Thank you so much!

tcarmeli1 commented on Packs/WALLIXBastion/Integrations/WAB/WAB.yml:
```diff
@@ -3326,18 +4394,51 @@ script:
+
+      category: Ldap Mappings.'
+    arguments:
+    - name: group_by
```
you can use here predefined :)

yoennb replied:
done

tcarmeli1 commented on Packs/WALLIXBastion/Integrations/WAB/WAB.yml:
```diff
@@ -3587,20 +4706,60 @@ script:
       required: true
+      auto: PREDEFINED
+      predefined:
+      - email
```
if email is the only option- why do we need this argument?

yoennb replied:
The API was designed with the intent that the list of notification types would grow in the future

tcarmeli1 commented on Packs/WALLIXBastion/Integrations/WAB/WAB.py:
```diff
@@ -2419,11 +3521,15 @@ def main() -> None:
-        deprecated = {
-            "wab-get-metadata-of-one-or-multiple-sessions": client.get_session_metadata,
-        }
+        deprecated: Dict[str, Any] = {}
```
you can delete this one and if so, you can also delete the statment in line 3538: 
        elif command in deprecated:
            LOG(f"WARNING: use of deprecated command {command}")
            return_results(deprecated[command](args))

yoennb replied:
done

tcarmeli1 commented on Packs/WALLIXBastion/ReleaseNotes/3_0_0.md:
```diff
@@ -0,0 +1,86 @@
+- **Breaking changes:** removed deprecated arguments in:
+    - wab-edit-application: application_put__meters
+
+- **Breaking changes:** removed deprecated outputs in:
```
please spilt so each remove in a command will be in a separate line
e.g., removed the XXX outputs in the YYY coammand.

yoennb replied:
done

tcarmeli1 reviewed (changes requested):
Hi @yoennb ,
Thank you for your contribution!
Great work :)
Please see my comments.
See you tomorrow in the Demo.
Thank you so much!

yoennb commented on Packs/WALLIXBastion/Integrations/WAB/WAB_description.md:
```diff
@@ -1,21 +1,31 @@
 
 Go to `Configuration` > `API Keys` to create one.
 
+> Note: using an API key rather than a password allows unrestricted aministrative access to the Bastion.
```
fixed

yoennb commented on Packs/WALLIXBastion/ReleaseNotes/3_0_0.md:
```diff
@@ -0,0 +1,86 @@
+    - add optional parameters service_post_subprotocols
+    - add service_put_seamless_connection option for RAWTCPIP
+
+- wab-edit-application: add following optional parameters:
```
done

yoennb commented on Packs/WALLIXBastion/ReleaseNotes/3_0_0.md:
```diff
@@ -0,0 +1,86 @@
+    - set service_post_subprotocols as non required 
+    - add service_post_seamless_connection option for RAWTCPIP
+
+- wab-edit-service-of-device: 
```
done

yoennb commented on Packs/WALLIXBastion/ReleaseNotes/3_0_0.md:
```diff
@@ -0,0 +1,86 @@
+    - wab-get-timeframes
+    - wab-get-user-group-restrictions
+
+- wab-add-service-in-device:
```
done

tcarmeli1 reviewed (changes requested):
Hi @yoennb ,
Just watched the demo!
Amazing work 👍 
Just a small comment about using predefined values: if the argument can have more than one value, it’s better to remove the predefined values, list the optional values in the description, and specify that a comma-separated list can be used. For example, in the video where you demonstrated adding days to the timeframe, it would be better not to use predefined values but instead specify in the description: 'Comma-separated list containing days. Optional values are: Sunday, Monday, Tuesday, Wednesday, Thursday, Friday, Saturday.' You can apply this template for all changes related to this comment.
Thank you so much!
Feel free to contact me :)

yoennb commented on Packs/WALLIXBastion/Integrations/WAB/WAB.py:
```diff
@@ -11,28 +11,60 @@ class AuthError(Exception):
-def to_markdown(name, t):
+def to_markdown(name: str, t):
     try:
         return tableToMarkdown(name, t)
```
Thank you for the information. I think that it's fine, it may help the user understand the full data structure he is working on. Most of the time, the Bastion does not return unset fields, so the presence of a null value is likely a kind of information.

tcarmeli1 reviewed (approved):
Great work!
